factcast-snapshotcache-mongodb/src/main/java/org/factcast/core/snap/mongo/MongoDbSnapshotCache.java

```diff
@@ -0,0 +1,151 @@
+
+    // Second index is for the projection class only, which is used for snapshots that are not
+    // related to an aggregate
+    String aggregateIdIndexResult =
+        collection.createIndex(Indexes.ascending(PROJECTION_CLASS_FIELD));
+    log.debug(
+        "Create aggregate ID index on factus_snapshot collection returned: {}",
+        aggregateIdIndexResult);
+
+    // Third index for TTL management of documents, expires the document after 0 seconds of the
+    // `expireAt` field. So, immediately.
+    collection.createIndex(
+        Indexes.ascending(EXPIRE_AT_FIELD), new IndexOptions().expireAfter(0L, TimeUnit.SECONDS));
+  }
+
+  @Override
+  public @NonNull Optional<SnapshotData> find(@NonNull SnapshotIdentifier id) {
+    Document query = getDocumentById(id);
+
+    Document result = collection.find(query).first();
+
+    if (result != null) {
+      Binary binary = result.get(SERIALIZED_PROJECTION_FIELD, Binary.class);
+      byte[] bytes = binary.getData();
+      String serializerId = result.getString(SNAPSHOT_SERIALIZER_ID_FIELD);
+      UUID lastFactId = UUID.fromString(result.getString(LAST_FACT_ID_FIELD));
+
+      tryUpdateExpirationDate(id);
+
+      return Optional.of(
+          new SnapshotData(bytes, SnapshotSerializerId.of(serializerId), lastFactId));
+    }
+
+    return Optional.empty();
+  }
+
+  private void tryUpdateExpirationDate(SnapshotIdentifier id) {
+    try {
+      collection.updateOne(
+          getDocumentById(id),
+          Updates.set(
+              EXPIRE_AT_FIELD,
+              Instant.now().plus(properties.getDeleteSnapshotStaleForDays(), ChronoUnit.DAYS)));
+    } catch (Exception e) {
+      log.warn("Failed to update expiration date for snapshot with id: {}", id, e);
+    }
+  }
+
+  @Override
+  public void store(@NonNull SnapshotIdentifier id, @NonNull SnapshotData snapshot) {
+    Binary bytes = new Binary(BsonBinarySubType.BINARY, snapshot.serializedProjection());
+
+    Document doc =
+        new Document(PROJECTION_CLASS_FIELD, id.projectionClass().getName())
+            .append(SNAPSHOT_SERIALIZER_ID_FIELD, snapshot.snapshotSerializerId().name())
+            .append(LAST_FACT_ID_FIELD, snapshot.lastFactId().toString())
+            // Document limit for binary data(16MB)
+            .append(SERIALIZED_PROJECTION_FIELD, bytes);
+
+    if (id.aggregateId() != null) {
+      doc.append(AGGREGATE_ID_FIELD, id.aggregateId().toString());
+    }
+
+    // Add expiration time to the document
+    doc.append(
+        EXPIRE_AT_FIELD,
+        Instant.now().plus(properties.getDeleteSnapshotStaleForDays(), ChronoUnit.DAYS));
+
+    Document query = getDocumentById(id);
+    collection.replaceOne(query, doc, new ReplaceOptions().upsert(true));
+  }
+
+  @Override
+  public void remove(@NonNull SnapshotIdentifier id) {
+    collection.deleteOne(getDocumentById(id));
+  }
+
+  private Document getDocumentById(SnapshotIdentifier id) {
```

Comment:
maybe the name is confusing. getDocumentById passing an id returning a document needs me to look twice in oder to see it was not the actual document returned, but a document containing the query :)

createByIdQuery(id) ?